batchlib/analysis/cell_level_analysis.py

```diff
@@ -199,109 +182,178 @@ def denoise(self, img):
-            serum -= (serum[cell_seg == ignore_label]).mean()
-
-        def get_per_mask_statistics(data):
-            per_cell_values = [data[cell_seg == label] for label in labels]
-            sums = data.new([arr.sum() for arr in per_cell_values])
-            means = data.new([arr.mean() for arr in per_cell_values])
-            instance_sizes = data.new([len(arr.view(-1)) for arr in per_cell_values])
-            top50 = np.array([0 if len(t) < 50 else t.topk(50)[0][-1].item()
-                              for t in per_cell_values])
-            top30 = np.array([0 if len(t) < 30 else t.topk(30)[0][-1].item()
-                              for t in per_cell_values])
-            top10 = np.array([0 if len(t) < 10 else t.topk(10)[0][-1].item()
-                              for t in per_cell_values])
-            # convert to numpy here
-            return dict(sums=sums.cpu().numpy(),
-                        means=means.cpu().numpy(),
-                        sizes=instance_sizes.cpu().numpy(),
-                        top50=top50,
-                        top30=top30,
-                        top10=top10)
+            for channel in channels:
+                channel -= (channel[cell_seg == ignore_label]).mean()
 
         cell_properties = dict()
-        cell_properties['marker'] = get_per_mask_statistics(marker)
-        cell_properties['serum'] = get_per_mask_statistics(serum)
+        for key, channel in zip(self.channel_keys, channels):
+            cell_properties[key] = self.get_per_instance_statistics(channel, cell_seg, labels)
         cell_properties['labels'] = labels.cpu().numpy()
 
         return cell_properties
 
     # this is what should be run for each h5 file
     def save_all_stats(self, in_file, out_file, device):
-        infected_threshold = self.infected_threshold  # TODO put this and other params into args of __init__
-        split_statistic = self.split_statistic
         sample = self.load_sample(in_file, device=device)
-        per_cell_statistics_to_save = self.eval_cells(*sample)
+        per_cell_statistics = self.eval_cells(*sample)
+
+        with open(out_file, 'wb') as f:
+            pickle.dump(per_cell_statistics, f)
 
-        per_cell_statistics = substract_background_of_marker(per_cell_statistics_to_save)
+    def run(self, input_files, output_files, gpu_id=None):
+        with torch.no_grad():
+            if gpu_id is not None:
+                os.environ['CUDA_VISIBLE_DEVICES'] = str(gpu_id)
+                device = torch.device(0)
+            else:
+                device = torch.device('cpu')
+            _save_all_stats = partial(self.save_all_stats, device=device)
+            for input_file, output_file in tqdm(list(zip(input_files, output_files)),
+                                                desc='extracting cell-level features'):
+                _save_all_stats(input_file, output_file)
+
+
+class CellLevelAnalysis(BatchJobWithSubfolder):
```

Comment:
Again, would rather use `BatchJobOnContainer` and then write to a hdf5 table.

Reply:
Yep, will be a one-row table at `measures_identifier` or something